libraries/ESP8266WiFi/src/ESP8266WiFiMulti.cpp

```diff
@@ -38,8 +38,104 @@ bool ESP8266WiFiMulti::addAP(const char* ssid, const char *passphrase) {
+            DEBUG_WIFI_MULTI("[WIFI] SSID: %s\n", WiFi.SSID().c_str());
+            DEBUG_WIFI_MULTI("[WIFI] IP: %d.%d.%d.%d\n", ip[0], ip[1], ip[2], ip[3]);
+            DEBUG_WIFI_MULTI("[WIFI] MAC: %02X:%02X:%02X:%02X:%02X:%02X\n", mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
+            DEBUG_WIFI_MULTI("[WIFI] Channel: %d\n", WiFi.channel());
+            break;
+        case WL_NO_SSID_AVAIL:
+            DEBUG_WIFI_MULTI("[WIFI] Connecting Failed AP not found.\n");
+            break;
+        case WL_CONNECT_FAILED:
+            DEBUG_WIFI_MULTI("[WIFI] Connecting Failed.\n");
+            break;
+        default:
+            DEBUG_WIFI_MULTI("[WIFI] Connecting Failed (%d).\n", status);
+            break;
+    }
+#else
+    (void)status;
+#endif
+}
+
+struct BestNetwork {
+    WifiAPEntry Network;
+    int         NetworkDb;
+    uint8       BSSID[6];
+    int32_t     Channel;
+
+    BestNetwork() : Network({NULL, NULL}), NetworkDb(INT_MIN) {}
+};
+
+void ESP8266WiFiMulti::selectBest(BestNetwork& best, int8_t scanResult) {
+    for(int8_t i = 0; i < scanResult; ++i) {
+
+        String ssid_scan;
+        int32_t rssi_scan;
+        uint8_t sec_scan;
+        uint8_t* BSSID_scan;
+        int32_t chan_scan;
+        bool hidden_scan;
+
+        WiFi.getNetworkInfo(i, ssid_scan, sec_scan, rssi_scan, BSSID_scan, chan_scan, hidden_scan);
+
+        bool known = false;
+        for(auto entry : APlist) {
+            if(ssid_scan == entry.ssid) { // SSID match
+                known = true;
+                if(rssi_scan > best.NetworkDb) { // best network
+                    if(sec_scan == ENC_TYPE_NONE || entry.passphrase) { // check for passphrase if not open wlan
+                        best.NetworkDb = rssi_scan;
+                        best.Channel = chan_scan;
+                        best.Network = entry;
+                        memcpy((void*) &best.BSSID, (void*) BSSID_scan, sizeof(best.BSSID));
+                    }
+                }
+                break;
+            }
+        }
+
+        if(known) {
+            DEBUG_WIFI_MULTI(" ---> ");
+        } else {
+            DEBUG_WIFI_MULTI("      ");
+        }
+
+        DEBUG_WIFI_MULTI(" %d: [%d][%02X:%02X:%02X:%02X:%02X:%02X] %s (%d) %c\n", i, chan_scan, BSSID_scan[0], BSSID_scan[1], BSSID_scan[2], BSSID_scan[3], BSSID_scan[4], BSSID_scan[5], ssid_scan.c_str(), rssi_scan, (sec_scan == ENC_TYPE_NONE) ? ' ' : '*');
+        delay(0);
+    }
+}
+
+wl_status_t ESP8266WiFiMulti::connectLoop() {
+    wl_status_t status = WiFi.status();
+    // wait for connection, fail, or timeout
+    if (status == WL_CONNECTED      ||
+        status == WL_NO_SSID_AVAIL  ||
+        status == WL_CONNECT_FAILED ||
+        (millis() > connect_timeout_ms))
+    {
+        handleConnectComplete(status);
+        connect_timeout_ms = 0;
+    }
+
+    return status;
+}
+
 wl_status_t ESP8266WiFiMulti::run(void) {
 
+    if (connect_timeout_ms != 0) {

```

Comment:
The timeout value is cleared above and never set again. Is that correct behavior?